src/crypto.rs

```diff
@@ -0,0 +1,206 @@
+    {
+        context_get_property_multiple(self, types::Property::CcmTag)
+    }
+}
+
+/// Implementation of all GCM/CCM properties for decrypt/open
+pub trait ContextWithXcmDecryptProperties: Context {
+    /// Sets the GCM AAD property.
+    fn set_property_gcm_aad(&self, gcm_aad: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmAad, gcm_aad)
+    }
+    /// Sets the GCM tag property.
+    fn set_property_gcm_tag(&self, gcm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmTag, gcm_tag)
+    }
+    /// Sets the CCM AAD property.
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>;
+    // {
+    //     OPERATION_set_input_length(self, input_len)?;
+    //     crypto::context_set_property_multiple(ctx, types::Property::CcmAad, ccm_aad)
+    // }
+    /// Sets the CCM tag property.
+    fn set_property_ccm_tag(&self, ccm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::CcmTag, ccm_tag)
+    }
+}
+
+
+/// Initializes the library. Must be called before any other crypto
+/// function. Should be called once in each thread that uses yaca.
+pub fn initialize() -> Result<()>
+{
+    let r = unsafe {
+        lib::yaca_initialize()
+    };
+    conv::res_c_to_rs(r)
+}
+
+/// Cleans up the library.
+/// Must be called before exiting the thread that called yaca_initialize().
+pub fn cleanup()
+{
+    unsafe {
+        lib::yaca_cleanup()
+    }
+}
+
+/// Safely compares first length bytes of two buffers.
+pub fn memcmp(first: &[u8], second: &[u8], length: usize) -> Result<bool>
+{
+    let min = std::cmp::min(first.len(), second.len());
+    assert!(length <= min);
+    let first = first.as_ptr() as *const c_void;
+    let second = second.as_ptr() as *const c_void;
+    let r = unsafe {
+        lib::yaca_memcmp(first, second, length)
+    };
+    conv::res_c_to_rs_bool(r)
+}
+
+/// Generates random data.
+pub fn random_bytes(length: usize) -> Result<Vec<u8>>
+{
+    let mut v: Vec<u8> = Vec::with_capacity(length);
+    let data = v.as_mut_ptr() as *mut c_char;
+    let r = unsafe {
+        lib::yaca_randomize_bytes(data, length)
+    };
+    conv::res_c_to_rs(r)?;
+    unsafe {
+        v.set_len(length);
+    };
+    Ok(v)
+}
+
+fn context_set_property_single<T, U>(ctx: &T, prop: types::Property,
+                                     value: U) -> Result<()>
+    where T: Context + ?Sized,
+{
+    let ctx = ctx.get_handle();
+    let property = conv::property_rs_to_c(&prop);
+    let value: *const U = &value;
+    let value = value as *const c_void;
+    let value_len = mem::size_of::<U>();
+    let r = unsafe {
+        lib::yaca_context_set_property(ctx, property, value, value_len)
+    };
+    conv::res_c_to_rs(r)
+}
+
+// Used by set_property_ccm_aad implementators
+pub(crate) fn context_set_property_multiple<T, U>(ctx: &T, prop: types::Property,
```

Comment:
This is a private function with a sole purpose of wrapping C code. I call it myself internally with proper types. I see no reason to change anything here.

Reply:
Should you decide to keep it, do the ```u8``` thing at least.

Reply:
The C function could in theory be used to pass &[size_t]. And the sizeof of single element would be 4.

Reply:
Such a use case does not currently arise in ```yaca-rust```, only ```u8``` is used.

Reply:
I don't care about current use cases. This is a wrapper around C function that was written to handle generic cases. I want this wrapper to be generic in the same way.

Reply:
Marks it ```unsafe``` then, like in https://github.com/Havner/yaca-rust/pull/1#discussion_r510988819

Reply:
> Coalescing both set_property functions into a single one with a match inside.
I could go for that, but how to do that with a single function?

Reply:
Also would bounding U to Integer be safe for you? As you suggested in the single variant.

Reply:
In light of recent discussions, I think that approach is out of the window.

Regardless, my idea was to make ```prop``` an enum encapsulating all possible property tags and their associated data. Hence the match (on ```prop```, to deduce c enum value, the pointer and data size).

Reply:
> In light of recent discussions, I think that approach is out of the window.

Why? Integer would cover all current and future use cases of those functions in YACA.

Reply:
That's right, I assumed you wanted to keep ```U``` unbound. My bad.

An integer bound on ```U``` would make both of these functions memory safe so that's bonus points.

As for squeezing it into a single function.. Not sure if it makes sense but maybe some data enum with two variants (one raw value or a slice)? Or just keep the slice one and pass single values as one-element slices?

Reply:
Ok, I've just bound U in all 3 functions to num::Integer.
I'll leave this unresolved for now and rethink making one function. I initially (when implementing this for the first time) wanted to use enums to encapsulate the data, but given set/get mechanism I could not figure out how to do that without creating 2 very similar enums, one with the data, one without. But as those enums are not public any more, this is acceptable. Still not sure it's worth it now, when this is not a public API.

Reply:
Agreed.

Reply:
I'll leave them as they are, consistent with other low level wrappers.